chore: Squash migration history and remove south compat

markstory opened the pull request:
This is a potentially breaking change for self-hosted instances as they will need to update to 2021-06 before moving forward.

Squashing migrations helps speed up new environment setup and CI when migrations are changed.

markstory committed:
chore: Squash migration history and remove south compat

This is a potentially breaking change for self-hosted instances as they
will need to update to 2021-06 before moving forward.

Squashing migrations helps speed up new environment setup and CI when
migrations are changed.

chadwhitacre commented:
So squashing adds a migration file but doesn't remove any? Then what's the point? 🤔

wedamija commented:
I tried running `makemigrations` with this in place and it created migration `0207` with no relation to the squash. Is there more we need to do here to make this squashed migration the 'primary' path or w/e?

markstory commented:
> So squashing adds a migration file but doesn't remove any? Then what's the point? 🤔

You can remove the migrations, but Django's docs say you should have the old migrations and the squash present in the tree until all environments are synced up.

Once that is done you can update the squash, remove the `replaces` attributes and delete all the old migrations.  My thinking was that we would have the replaced migrations until the July release for self-hosted is cut afterwhich I would remove the old migration files.

markstory committed:
Remove intermediary state churn.

Simplify migrations by remove state/operations that is reverted within
the same migration. Django doesn't fully handle separate state/database
well in big squash migrations. Eliding those operations makes migrations
run.

markstory committed:
Merge branch 'master' into chore/migration-squash

markstory commented:
@chadwhitacre @wedamija I think this is ready now. The migration SQL generation builds are failing but I think that is due to bugs in django migrations. 

I've been testing this locally using `make reset-db` and also running tests with `MIGRATIONS_TEST_MIGRATE=1 pytest tests/sentry/api/endpoints/test_organization*`

BYK reviewed (changes requested):
@markstory, can you add a changelog entry for this, telling people that they need to stop at this revision? @wedamija was able to cook in a warning telling people to upgrade to `9.1.2` first somehow, maybe we can also get that?

markstory committed:
Add release notes for migration upgrade.

markstory committed:
Update upgrade note.

markstory commented:
> can you add a changelog entry for this, telling people that they need to stop at this revision?

I've updated the changelog with a note.

wedamija reviewed (approved):
@BYK was asking whether we could bake in a warning to help prevent users shooting themselves in the foot here. One option could be to repurpose the south compatibility layer to make sure that people either have 0 migrations run, or up to migration 200 run and error otherwise.

Maybe we could do that when we delete the unsquashed migrations from the codebase.

markstory committed:
Merge branch 'master' into chore/migration-squash

getsantry[bot] committed:
style(lint): Auto commit lint changes

BYK commented on CHANGES:
```diff
@@ -10,6 +10,13 @@
 
 There will be a future change to where the frontend bundle will be loaded asynchronously. This will be a breaking change that can affect custom plugins that access certain globals in the django template. Please see https://forum.sentry.io/t/breaking-frontend-changes-for-custom-plugins/14184 for more information.
 
+21.7.1
+------
+
+- If you are upgrading from a release prior to 10.0.0 you need to first upgrade
+  to 21.6.1 and run migrations. In 21.7.1 the ability to upgrade from south based
+  migrations was removed.
```
This part needs to be moved to top, before `21.6.1` heading. I'd also add a `### Breaking Change` sub-heading like the one in `21.6.0` and put a link to https://develop.sentry.dev/self-hosted/#hard-stops

BYK reviewed (changes requested):
Requesting changes for the readme stuff. I'd also love to find a way to implement what @wedamija mentioned regarding in-code "into your face" error/warnings but I won't block on that.

markstory committed:
Clean up change log

BYK reviewed (approved):
Unblocking
